src/main/java/org/folio/rest/impl/CheckinHelper.java

```diff
@@ -47,19 +49,48 @@
 
     // 1. Get piece records from storage
     return retrievePieceRecords(checkinPieces)
-      // 2. Update items in the Inventory if required
+      // 2. Check locationId presence for pieces related to
+      // POLine with createInventory "Instance, Holding" or "Instance, Holding, Item"
+      .thenCompose(this::checkLocationId)
+      // 3. Update items in the Inventory if required
       .thenCompose(this::updateInventoryItems)
-      // 3. Update piece records with checkIn details which do not have
+      // 4. Update piece records with checkIn details which do not have
       // associated item
       .thenApply(this::updatePieceRecordsWithoutItems)
-      // 4. Update received piece records in the storage
+      // 5. Update received piece records in the storage
       .thenCompose(this::storeUpdatedPieceRecords)
-      // 5. Update PO Line status
+      // 6. Update PO Line status
       .thenCompose(this::updatePoLinesStatus)
-      // 6. Return results to the client
+      // 7. Return results to the client
       .thenApply(piecesGroupedByPoLine -> prepareResponseBody(checkinCollection, piecesGroupedByPoLine));
   }
 
+  /**
+   * Checks locationId presence for pieces related to POLine with
+   * createInventory = "Instance, Holding" or "Instance, Holding, Item".
+   *
+   * @return {@link CompletableFuture} which holds map with PO line id as key
+   *         and list of corresponding pieces as value
+   */
+  private CompletableFuture<Map<String, List<Piece>>> checkLocationId(Map<String, List<Piece>> piecesRecords) {
+    getPoLines(piecesRecords)
+      .thenAccept(poLinesReply -> {
+        for (PoLine poLine : poLinesReply) {
+          Eresource.CreateInventory eci = poLine.getEresource() != null ? poLine.getEresource().getCreateInventory() : null;
+          Physical.CreateInventory pci = poLine.getPhysical() != null ? poLine.getPhysical().getCreateInventory() : null;
+          if (Eresource.CreateInventory.INSTANCE_HOLDING.equals(eci) || Eresource.CreateInventory.INSTANCE_HOLDING_ITEM.equals(eci)
+            || Physical.CreateInventory.INSTANCE_HOLDING.equals(pci) || Physical.CreateInventory.INSTANCE_HOLDING_ITEM.equals(pci)) {
+            for (CheckInPiece cip : checkinPieces.get(poLine.getId()).values()) {
+              if (cip.getLocationId() == null) {
```

Comment:
1. Seems "P/E Mix" is not considered here. For such PO Lines E- and Physical resources are fully independent. To validate properly records for such PO Line piece format needs to be taken into account as well.
2. When looking at `updatePieceWithCheckinInfo` method here if location is not provided in the request for particular piece, it remains unchanged:
```java
    if (StringUtils.isNotEmpty(checkinPiece.getLocationId())) {
      piece.setLocationId(checkinPiece.getLocationId());
    }
```
So the entry in the request like following
```json
{
  "id": "71d9322b-5cdd-45d8-ad45-c7f3044802e7",
  "itemStatus":"Received"
}
```
means leave location unchanged but just check-in piece. So validation needs to be a bit enhanced.
@folio-org/acquisitions - please have a look if you agree with such approach.

Reply:
+1 wrt handling of P/E Mix.

> if location is not provided in the request for particular piece, it remains unchanged
I don't think this is the way it should work.  In cases where location is required (based on createInventory), location must be specified.  If not, an error should be returned.

Reply:
Current approach for check-in/receiving flows contains 2 steps:

1. Create piece (and item).
   - For check-in an user can specify location along with other data (caption, comment, itemId etc.).
   - For receiving flow the location will be always specified on Order transition to Open.
2. Check-in/receive that piece. At this stage location might be already specified in piece record. That is why location might be omitted in the request

UI works with the piece data to preselect location in the DDL and an user can change it when checking-in/receiving a piece. So it is expected that the location actually will be forced to be provided by UI anyway.
Forcing `locationId` in the check-in/receiving request (if `createInventory` of the PO Line requires is) by business logic is a bit inconsistent with other properties like `comment` or `caption`. But taking into account that at the moment the only consumer of the receiving/check-in API is UI app, the business logic mostly is based on what UI sends...

Reply:
> I don't think this is the way it should work. In cases where location is required (based on createInventory), location must be specified. If not, an error should be returned.

Current implementation works without piece locationId validation, i.e. error will be returned if only CheckIn request doesn't contain info about locationId.